auth/auth/flow.py

```diff
@@ -0,0 +1,123 @@
+import abc
+from typing import Optional
+import msal
+import google.auth.transport.requests
+import google.oauth2.id_token
+import google_auth_oauthlib.flow
+import aiohttp.web
+import json
+import urllib.parse
+
+from gear.cloud_config import get_global_config
+
+
+class FlowData:
+    def __init__(self, authorization_url: str, state: str):
+        self.authorization_url = authorization_url
+        self.state = state
+
+
+class FlowResult:
+    def __init__(self, login_id: str, email: str, token: dict):
+        self.login_id = login_id
+        self.email = email
+        self.token = token
+
+
+class Flow(abc.ABC):
```

Comment:
OK, I've thought about this a bit more and I think we can streamline these classes a bit more.

Before this PR a login session had a `state` parameter. After this PR a login session will have a `state` parameter and a `flow` parameter. The `flow` parameter is only used by Azure. The value of the `state` parameter in Azure is duplicated in the `flow` parameter. My conclusion: we should just have one `flow` session parameter which is a dictionary. In Google, it will have one field: `state`. In Azure, it will have whatever is returned by `initiate_auth_code_flow`. This removes a few lines across the code base (woo!).

Below, I note that in Azure we unnecessarily initiate the auth flow the second time we call `get_flow` (the call to `get_flow` in oauth2callback). I think this mistake was made easier by the structure of this class. Let's instead make the structure of the class mirror the steps of the oauth2 flow:
```python3
class Flow(abc.ABC):
    @abc.abstractmethod
    def initiate_flow(redirect_uri: str) -> dict:
        """Initiates the OAuth2 flow. Usually run in response to a user clicking a login button.

           The returned dict should be stored in a secure session so that the server can
           identify to which OAuth2 flow a client is responding. In particular, the server must
           pass this dict to :meth:`.receive_callback` in the OAuth2 callback."""
        raise NotImplementedError

    @abc.abstractmethod
    def receive_callback(self, flow_dict: dict) -> FlowResult:
        """Concludes the OAuth2 flow by returning the user's identity and credentials."""
        raise NotImplementedError
```

Note that there should be exactly one `Flow` class for the entire lifetime of the server. It will need, I think, just one field: `credentials_file`.

In auth.py, I think this produces substantial simplifications, for example login and signup both only need one line for OAuth2 related activity:
```python3
sessoin['flow'] = app['the_oauth2_flow'].initiate_flow(deploy_config.external_url('auth', '/oauth2callback'))
```
And oauth2callback is likewise simplified to two lines:
```python3
flow_result = app['the_oauth2_flow'].receive_callback(session['flow'])
login_id = flow_result.login_id
```